### ISSUE TITLE
[naming-convention] Missing distinction between "declaring" versus "using" a property

### ISSUE THREAD

**Issue by octogonz** [typescript-eslint/typescript-eslint#2244]
Regarding the `property` selector, the [naming-convention docs](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/naming-convention.md) say:

> - `property` - matches any object, class, or object type property. Does not match properties that have direct function expression or arrow function expression values.
>  - Allowed `modifiers`: `private`, `protected`, `public`, `static`, `readonly`, `abstract`.
>  - Allowed `types`: `boolean`, `string`, `number`, `function`, `array`.

In other words, there is no distinction between an `interface` property versus an object property.  Consider this example:

```ts
interface IMyContract {
  // We very much need ESLint to forbid this:
  WrongCapitalization: string;
}

ancientLegacyThing.doSomething({
  // But we have absolutely no control over this contract,
  // so ESLint should leave it alone:
  AlsoWrongCapitalization: 'hello'
});
```

**Expected Result**

The `naming-convention` rule should provide selectors that distinguish between declaring an "interface", versus making an "object" that will fulfill an existing interface.

**Actual Result**

As far as I can tell, there is no distinction.  In a code base with lots of legacy APIs, it would be completely impractical to add a lint suppression for every such function call.

As a result, we basically have to disable the `naming-convention` validation for any selector that could potentially be used to fulfill a legacy contract.  This seems like a fairly severe limitation.

**Additional Info**

`@typescript-eslint/naming-convention` replaces 4 earlier rules which were recently deleted (which is how I got involved with this all the sudden):

- `@typescript-eslint/camelcase`
- `@typescript-eslint/class-name-casing`
- `@typescript-eslint/interface-name-prefix`
- `@typescript-eslint/member-naming`

I am trying to understand why those earlier rules did not seem to exhibit this problem.  Maybe they actually were not validating interface property names?

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.3.0` |
| `@typescript-eslint/parser`        | `3.3.0` |
| `TypeScript`                       | `3.5.3` |
| `ESLint`                           | `7.2.0` |
| `node`                             | `12.17.0` |
| `npm`                              | `6.14.4` |

**Comment by octogonz** [comment]
The `property` selector also seems to be used for destructuring assignments:

```ts
// Why does CONSTANT1 match the "property" selector? 
// It is a "variable" in this context.
const { CONSTANT1, CONSTANT2 } = ancientLegacyThing.constants;
```

**Comment by bradzacher** [comment]
> Why does CONSTANT1 match the "property" selector? 

I believe it'll actually try to match both the variable and property selectors.
This is a bug. If you wouldn't mind filing a new issue for this.

The bug occurs because `Property` is reused for both object literals and object destructuring, and the code doesn't ignore properties in a destructuring context.

This should pull the identifiers from the destructuring and analyse them as a variable
https://github.com/typescript-eslint/typescript-eslint/blob/5fdd21a1726fb6928098c4152aec55a30df960d4/packages/eslint-plugin/src/rules/naming-convention.ts#L433-L445
	
And this should pull the identifiers as a property
https://github.com/typescript-eslint/typescript-eslint/blob/5fdd21a1726fb6928098c4152aec55a30df960d4/packages/eslint-plugin/src/rules/naming-convention.ts#L518-L523

----

> Maybe they actually were not validating interface property names?

Our extension rule simply [lumped interface properties in with normal properties](https://github.com/typescript-eslint/typescript-eslint/blob/v2.34.0/packages/eslint-plugin/src/rules/camelcase.ts#L130-L137).

So depending on your [configuration for `camelcase`](https://eslint.org/docs/rules/camelcase#options), it may not have validated property names, or destructuring.

----

There could be an additional selector for "property signatures" to distinguish between type declarations and object literals.

Though at first glance, this doesn't seem like a great idea, as this would make it easy to configure a naming convention that is impossible to satisfy. I.e. "typeProperty format is snake_case" and "objectProperty format is camelCase" together would mean you can never declare an object literal against a known object type.

----

The problem here is "what constitutes usage?"

Is it just an object literal inline in a call that constitutes usage?
What about other cases?
- variable declaration without a type annotation `const x = {a: 1}`
- variable declaration with a type annotation `const x: Foo = {a: 1}`
- type assertion `({a: 1} as Foo)`
- return statement `return {a: 1}`
- variable declaration without a type annotation that's later used in one of the above cases `const x = {a : 1}; const y: Foo = x`

are all of these cases "usages"? Just some of them?
It's hard to define this exactly, and likely whatever definition you arrive at would _feel_ kind of broken, as the rule would ignore some properties but not others.

The best way to do this would be to use type-information to analyse the context of the property to determine if it is a "freehand" or a "predefined" property name, but that's going to be reasonably expensive to do, and would be a breaking change for the rule.

Maybe this could be added as a new selector for people that have a mix-and-match of "properly defined" and legacy/externally controlled properties?

Unsure if this is something people will find value in, as I don't have any idea about how common these legacy/externally controlled properties are in the real world.

**Comment by octogonz** [comment]
> > Why does CONSTANT1 match the "property" selector?
> 
> I believe it'll actually try to match both the variable and property selectors.
> This is a bug. If you wouldn't mind filing a new issue for this.

Here you go: https://github.com/typescript-eslint/typescript-eslint/issues/2250

> There could be an additional selector for "property signatures" to distinguish between type declarations and object literals.
> 
> Though at first glance, this doesn't seem like a great idea, as this would make it easy to configure a naming convention that is impossible to satisfy. I.e. "typeProperty format is snake_case" and "objectProperty format is camelCase" together would mean you can never declare an object literal against a known object type.

This is maybe an unavoidable reality of the language, that we cannot escape. 

For what it's worth, my frustration in this exercise was with not being able to express what I wanted. I didn't mind having to debug all the interactions, or carefully think through all the cases. Most of those issues come to light pretty quickly when you test your ruleset on a huge code base.

In fact, I ended up having to make [a macro](https://github.com/microsoft/rushstack/blob/1b530bb644d2fa89212ec322a3ce79f17e6abd5b/stack/eslint-config/macros.js#L63) to generate the `naming-convention` configuration. I had to consider [each single selector individually](https://github.com/microsoft/rushstack/blob/1b530bb644d2fa89212ec322a3ce79f17e6abd5b/stack/eslint-config/index.js#L161).  The "grouping selectors" only seemed to muddy the waters.  This might not be representative of a casual user, though. 🙂

What I will say is that the end result was still superior to forking **naming-convention.ts** (which I seriously considered heheh).  Your all-in-one declarative approach does seem like the right design.

> The problem here is "what constitutes usage?"
> 
> Is it just an object literal inline in a call that constitutes usage?
> What about other cases?
> 
>     * variable declaration without a type annotation `const x = {a: 1}`

We don't want to lint `{ a: 1 }` because `@typescript-eslint/typedef` already enforces naming via `x`.

>     * variable declaration with a type annotation `const x: Foo = {a: 1}`

We don't want to lint `{ a: 1 }` because `@typescript-eslint/typedef` already enforces naming via `Foo`

>     * type assertion `({a: 1} as Foo)`

We don't want to lint `{ a: 1 }` because `@typescript-eslint/typedef` already enforces naming via `Foo`

>     * return statement `return {a: 1}`

We don't want to lint `{ a: 1 }` because `@typescript-eslint/explicit-function-return-type` already does that.

>     * variable declaration without a type annotation that's later used in one of the above cases `const x = {a : 1}; const y: Foo = x`

Not allowed because of `@typescript-eslint/typedef`.

> are all of these cases "usages"? Just some of them?
> It's hard to define this exactly, and likely whatever definition you arrive at would _feel_ kind of broken, as the rule would ignore some properties but not others.

All are usages. Your view here seems biased by a coding style that relies heavily on type inference. It's a very popular style, particularly for smaller teams who optimize for "writing code really concisely." Larger teams often find themselves optimizing for "ensuring that an unknown person's code will be easy for me to review/maintain."

**Comment by bradzacher** [comment]
I think then the best course of action would be the following:

Define some new selectors:
- `propertySignature`, which only matches an object type property signatures.
  - for parity with `property`, this _should not_ match property signatures with a function type.
- `propertyDeclaration`, which only matches object literal's property declarations.
  - this selector would be a rename of the existing `property` selector, minus matching property signatures.
- `methodSignature`, which only matches an object type's method signatures.
  - for parity with `property`, this _should_ match property signatures with a function type.
- `methodDeclaration`, which only matches object literal's method declarations.
  - this selector would be a rename of the existing `method` selector, minus matching method signatures.

Define some new group selectors for backwards compatibility and easy of consumption:
- `property`, which groups `propertySignature` and `propertyDeclaration`.
- `method`, which groups `methodSignature` and `methodDeclaration`.

Also update the `memberLike` group selector to include all of the above selectors.

This would give you the flexibility to define a naming convention that ignores object literal properties altogether.

One caveat to note here is that without a convention defined for `*Declaration`, there will be no way to be able to save you from index types:

```ts
const x: {[k: string]: unknown} = {
 aaaaa: 1,
 BBBBB: 2,
 AbAbA: 3,
 AA_BB: 4,
 aa_bb: 5,
 // etc
}
```

**Comment by JoshuaKGoldberg** [comment]
Is `const { abc } = def` matching on both property and variable declarations still considered a bug? I can [send a PR](https://github.com/typescript-eslint/typescript-eslint/compare/master...JoshuaKGoldberg:naming-convention-object-patterns-variables-only?expand=1) if so.

**Comment by bradzacher** [comment]
![image](https://user-images.githubusercontent.com/7462525/93167184-ef821200-f6d4-11ea-8975-7cf52d04ee6a.png)

**Comment by ab-pm** [comment]
I'd love to see `naming-convention` not complain about any names that are declared elsewhere and are typechecked against some interface. There's so many third-party libraries expecting weird parameter, often when interfacing with different ecosystems. I wish the plugin could just ignore names over which I have no control anyway.

I admit this would be quite complicated. We can't just check whether a property is part of the type that the object has, as the type might be inferred from the very object literal itself (e.g. in `as const` declarations). It would actually need to find the declaration of that type, and complain only if there exists none.

**Comment by paulius-valiunas** [comment]
Any update on this? It's very annoying that I have to add `eslint-ignore` everywhere I use `fetch` and similar APIs that require non-camelCase arguments.

**Comment by bradzacher** [comment]
With any issue in opened in this project - it either has a visible progress in the form of an attached PR, or it has no progress.

We are a community run project. The **volunteer** maintainers spend most of their time triaging issues and reviewing PRs. This means that most issues *will not progress* unless a member of the community steps up and **champions** it.

If this issue is important to you - consider being that champion.

If not - please just subscribe to the issue and wait patiently.
Commenting asking for status updates does not bump issue priority in any way and just serves to spam ***everyone*** subscribed to the issue.

**Comment by lucasbasquerotto** [comment]
> The problem here is "what constitutes usage?"

In my opinion, I expect this linting rule to only validate cases which are explicitly defined by the user, basically:

- Variable definition

```ts
const correctCase = 1; // ok
const WrongCase = 2; // error
```

- Function parameters

```ts
const myFunction = (param1, WrongParam2) => { // error at WrongParam2
    // function body
}
```

- Interface definition:

```ts
interface MyContract {
    correctCase: string; // ok
    WrongCase: string; // error
    myFunction1: (param1, param2) => void; // ok
    myFunction2: (param1, WrongParam2) => void; // error at WrongParam2
}
```

It should validate only in the above cases (I would be ok with a flag to validate only in the cases above **only if the flag is enabled**, for backwards compatibility, to allow the lint to work as it's working today for the people that want the validation in every place).

The following cases wouldn't be validated, because those cases **could** be usage (and, in this case, it should not be a responsibility of the user the proper casing):

- variable declaration without a type annotation `const x = {a: 1}`
- variable declaration with a type annotation `const x: Foo = {a: 1}`
- type assertion `({a: 1} as Foo)`
- return statement `return {a: 1}`
- variable declaration without a type annotation that's later used in one of the above cases `const x = {a : 1}; const y: Foo = x`

For object destructuring, it should validate as variables (not as properties), which is the 1st case (variable definition):

```ts
// The below cases should validate only the variables names, not the properties names
const { CONSTANT1, CONSTANT2 } = someLibrary.constants; // error
const { CONSTANT1: constant1, CONSTANT2: constant2 } = someLibrary.constants; // ok
```

For now, I'm using broader rules (instead of stricter rules, as I wanted) because of this issue, otherwise I would have to add several comments to disable this rule in several lines along my code to avoid errors that are outside my control.

**Comment by paulius-valiunas** [comment]
@lucasbasquerotto I would validate "variable declaration without a type annotation" because by doing that you're _de facto_ defining an inline type. But even without this adjustment I still think your suggestion is great. Overall, problems should only be reported where the type is defined, not where it's used. That just creates a completely unnecessary snowball effect of lint errors coming from one type definition. For a lint rule, I think having false negatives is better than having false positives.

**Comment by bradzacher** [comment]
I agree with @paulius-valiunas - `const X = {a: 1};` should be validated as it has no type.

But `const x: Foo = {a: 1};` and `const x = {a: 1} as Foo;` shouldn't be validated because they have a type.

The difficult thing about this is "what if Foo is a "record" type - ` type Foo = Record<string, any>;`?"
Now you have a problem with properties that are not validated by the rule and are not covered by a type whose properties are validated by the rule. Eg now you have unchecked names!

This is unfortunately an impossible thing to solve without using type information. This sort of problem also explodes when you consider deeply nested objects whose parents have types.

There are other examples that can be complicated to handle even with types too:
```ts
declare function foo<T>(arg: T): T;
const x = foo({a: 1});
```
This usage looks like it should be contextually typed - it's an argument, which must have a type! However it is actually the same as the first variable declaration - meaning you again have an unchecked property name.

Another example would be return types.
`() => ({a: 1})` should be validated, similar to the first variable declaration it is untyped.
`(): Foo => ({a: 1})`. Functions and return statements do complicate things somewhat thanks to the layer of indirection they introduce.

This is why I say that it isn't a cut-and-dry issue. There's a lot more nuance to consider here than it might seem at first glance.

---

> variable declaration without a type annotation that's later used in one of the above cases `const x = {a : 1}; const y: Foo = x`

Whilst this might seem like a good idea - generally this kind of scope analysis usage creates a bad UX for the rule.

This is because it creates so-called "actions at a distance". Where the variable `x` will report in the lint rule except when used in a certain way. Also if you remove this usage, then it starts reporting where it didn't before. It can be very confusing!

Unless someone understands this system it looks like the rule has a false-negative bug. I talk about these a lot - but false negatives are bad because they harms user trust in the rule and the linting ecosystem.

**Comment by lucasbasquerotto** [comment]
@paulius-valiunas The problem with that case is that you don't really know if it's an independent type or if it isn't, because there are still several edge cases like @bradzacher pointed (although he agree with you in the `const X = {a: 1};` case).

Even if the linter has a way to know that the variable is being assinged to a typed variable (or being cast to a type), what if the property name is not in the type (like `type Foo = Record<string, any>`), in which case it would make sense, then, to validate.

It would make also sense to validate the argument passed to `function foo<T>(arg: T): T` too, unless it's changed to `function foo<T extends { a: string }>(arg: T): T`, for example.

And then the return types `() => ({a: 1})` and `(): Foo => ({a: 1})`. Only the first case should be validated... unless the result is being assigned to a typed variable or argument, and `a` is declared explicitly in the infered type.

The problem is that for the linter to be able to do that would be very complex, if possible at all, expecially considering indirect cases:

```ts
const foo = <T>(arg: T) => ({ result: arg });

const var1 = foo({ A: 1 }); // ok, you don't have control about the 'a' property name
callSomeExternalFunction1(var1);

const var2 = foo({ A: 1 }); // error, you have control about the 'a' property name
callSomeExternalFunction2(var2);

const var3 = { A: 1 }; // ok, for the same reason as var1 (changing to camelCase would give a type error)
const var4 = foo(var3);
callSomeExternalFunction1(var4); // but callSomeExternalFunction2 should give an error at 'const var3'

const var5 = foo({ A: 1 }); // error
console.log(var5);

// and then in some 3rd party library:

export const callSomeExternalFunction1 = (arg: { result: { A: number } }) => {
    // function body
}

export const callSomeExternalFunction2 = (arg: { result: Record<string, number> }) => {
    // function body
}
```

I don't know how the linter would be able to tell (or how complex it would be) that `const var1 = foo({ a: 1 })` is ok but `const var2 = foo({ a: 1 })` isn't. That's why I said (with emphasis in **could**):

> The following cases wouldn't be validated, because those cases **could** be usage

I'm not saying that I wouldn't like it to validate `const x = { A: 1 }` if the type is defined and used only by me, but that I don't know how the linter would know if the `A` property name is in my control or not, which could create false positives or negatives. Because of that, I considered better to consider those 3 cases, that are in your control:

- Variable definition
- Function declaration
- Interface definition (it might not be in your control if it's some object to be sent to an API that you don't have control, for example, but I think this case should be validated anyway)

**Comment by ab-pm** [comment]
> The difficult thing about this is "what if Foo is a "record" type - ` type Foo = Record<string, any>;`?" Now you have a problem with properties that are not validated by the rule and are not covered by a type whose properties are validated by the rule. Eg now you have unchecked names!

I would be totally fine with that for a `Record`. Using `string` means that any string is a valid property name - it's usually data - and I don't need a linter to have a word in that.

A more interesting question might be what should happen to
```ts
type Foo = Record<'A' | 'a' | '__a' | '_A_', any>;
```
Should the rule complain about these property names? Imo no, it should only complain about properties in `interface`s and literal types, and also no because this usage of names (probably) is too complicated to infer.

> ```ts
> const foo = <T>(arg: T) => ({ result: arg });
> 
> const var1 = foo({ A: 1 }); // ok, you don't have control about the 'a' property name
> callSomeExternalFunction1(var1);
> 
> const var3 = { A: 1 }; // ok, for the same reason as var1 (changing to camelCase would give a type error)
> const var4 = foo(var3);
> callSomeExternalFunction1(var4);
> ```

I disagree that these should be ok. The `var1`/`var3`/`var4` do get their types from the object literal definition, it is not inferred backwards from the `callSomeExternalFunction1` parameter - there is no contextual type for them. The property name should be linted to the naming convention.
To avoid that and get away with unconventional names, you should explicitly declare their type to be `Result` (that is hopefully exported from the 3rd party library - if not, `type Result = Parameters<typeof callSomeExternalFunction1>[0]['result']`).

I expect the following codes to pass linting:
```ts

**Comment by paulens12** [comment]
While I agree that false negatives reduce the trust in a rule, I think false positives are way worse than that. It's better to have a rule that covers at least _some_ of what you want to do than have to deal with a rule that keeps complaining about unimportant issues, or issues that just aren't in your control.

**Comment by paulius-valiunas** [comment]
While I agree that false negatives reduce the trust in a rule, I think false positives are way worse than that. It's better to have a rule that covers at least some of what you want to do than have to deal with a rule that keeps complaining about unimportant issues, or issues that just aren't in your control.

**Comment by JoshuaKGoldberg** [comment]
Per #8792, we're feature freezing this rule. It's reached the point of complexity where we're having a hard time justifying working on it. 

If you'd like to see this option happen, I'd encourage you to add it to a community project such as [eslint.style](https://eslint.style). Cheers! 💙
